# Project 07 -- Anthony Micco

**TA Help:** N/A
    
**Collaboration:** N/A

## Question 1

In [2]:
#initializing the database

In [3]:
%sql sqlite:////anvil/projects/tdm/data/lahman/lahman.db

In [4]:
# looking at the Teams table

In [5]:
%%sql
SELECT * FROM Teams LIMIT 10

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[(1871, 'NA', 'BS1', 'BNA', '', 3, 31, '', 20, 10, '', '', 'N', '', 401, 1372, 426, 70, 37, 3, 60, 19, 73, 16, '', '', 303, 109, 3.55, 22, 1, 3, 828, 367, 2, 42, 23, 243, 24, 0.834, 'Boston Red Stockings', 'South End Grounds I', '', 103, 98, 'BOS', 'BS1', 'BS1'),
 (1871, 'NA', 'CH1', 'CNA', '', 2, 28, '', 19, 9, '', '', 'N', '', 302, 1196, 323, 52, 21, 10, 60, 22, 69, 21, '', '', 241, 77, 2.76, 25, 0, 1, 753, 308, 6, 28, 22, 229, 16, 0.829, 'Chicago White Stockings', 'Union Base-Ball Grounds', '', 104, 102, 'CHI', 'CH1', 'CH1'),
 (1871, 'NA', 'CL1', 'CFC', '', 8, 29, '', 10, 19, '', '', 'N', '', 249, 1186, 328, 35, 40, 7, 26, 25, 18, 8, '', '', 341, 116, 4.11, 23, 0, 0, 762, 346, 13, 53, 34, 234, 15, 0.818, 'Cleveland Forest Citys', 'National Association Grounds', '', 96, 100, 'CLE', 'CL1', 'CL1'),
 (1871, 'NA', 'FW1', 'KEK', '', 7, 19, '', 7, 12, '', '', 'N', '', 137, 746, 178, 19, 8, 2, 33, 9, 16, 4, '', '', 243, 97, 5.17, 19, 1, 0, 507, 261, 5, 21, 17, 163, 8, 0.803, 'Fort Wayne Kekiongas', 'Hamilton Field', '', 101, 107, 'KEK', 'FW1', 'FW1'),
 (1871, 'NA', 'NY2', 'NNA', '', 5, 33, '', 16, 17, '', '', 'N', '', 302, 1404, 403, 43, 21, 1, 33, 15, 46, 15, '', '', 313, 121, 3.72, 32, 1, 0, 879, 373, 7, 42, 22, 235, 14, 0.84, 'New York Mutuals', 'Union Grounds (Brooklyn)', '', 90, 88, 'NYU', 'NY2', 'NY2'),
 (1871, 'NA', 'PH1', 'PNA', '', 1, 28, '', 21, 7, '', '', 'Y', '', 376, 1281, 410, 66, 27, 9, 46, 23, 56, 12, '', '', 266, 137, 4.95, 27, 0, 0, 747, 329, 3, 53, 16, 194, 13, 0.845, 'Philadelphia Athletics', 'Jefferson Street Grounds', '', 102, 98, 'ATH', 'PH1', 'PH1'),
 (1871, 'NA', 'RC1', 'ROK', '', 9, 25, '', 4, 21, '', '', 'N', '', 231, 1036, 274, 44, 25, 3, 38, 30, 53, 10, '', '', 287, 108, 4.3, 23, 1, 0, 678, 315, 3, 34, 16, 220, 14, 0.821, 'Rockford Forest Citys', 'Agricultural Society Fair Grounds', '', 97, 99, 'ROK', 'RC1', 'RC1'),
 (1871, 'NA', 'TRO', 'TRO', '', 6, 29, '', 13, 15, '', '', 'N', '', 351, 1248, 384, 51, 34, 6, 49, 19, 62, 24, '', '', 362, 153, 5.51, 28, 0, 0, 750, 431, 4, 75, 12, 198, 22, 0.845, 'Troy Haymakers', "Haymakers' Grounds", '', 101, 100, 'TRO', 'TRO', 'TRO'),
 (1871, 'NA', 'WS3', 'OLY', '', 4, 32, '', 15, 15, '', '', 'N', '', 310, 1353, 375, 54, 26, 6, 48, 13, 48, 13, '', '', 303, 137, 4.37, 32, 0, 0, 846, 371, 4, 45, 13, 218, 20, 0.85, 'Washington Olympics', 'Olympics Grounds', '', 94, 98, 'OLY', 'WS3', 'WS3'),
 (1872, 'NA', 'BL1', 'BLC', '', 2, 58, '', 35, 19, '', '', 'N', '', 617, 2571, 753, 106, 31, 14, 29, 28, 53, 18, '', '', 434, 166, 2.9, 48, 1, 1, 1548, 573, 3, 63, 77, 432, 22, 0.83, 'Baltimore Canaries', 'Newington Park', '', 106, 102, 'BAL', 'BL1', 'BL1')]

In [6]:
#join query to see the player's name, team name, year and average salary

In [7]:
%%sql
SELECT
    p.nameFirst || ' ' || p.nameLast AS fullName,
    t.name AS teamName,
    s.yearID,
    s.salary
FROM Salaries s
JOIN People p ON p.playerID = s.playerID
JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
ORDER BY s.salary DESC
LIMIT 10;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('Alex Rodriguez', 'New York Yankees', 2009, 33000000.0),
 ('Alex Rodriguez', 'New York Yankees', 2010, 33000000.0),
 ('Clayton Kershaw', 'Los Angeles Dodgers', 2016, 33000000.0),
 ('Clayton Kershaw', 'Los Angeles Dodgers', 2015, 32571000.0),
 ('Alex Rodriguez', 'New York Yankees', 2011, 32000000.0),
 ('Zack Greinke', 'Arizona Diamondbacks', 2016, 31799030.0),
 ('Alex Rodriguez', 'New York Yankees', 2012, 30000000.0),
 ('David Price', 'Boston Red Sox', 2016, 30000000.0),
 ('Alex Rodriguez', 'New York Yankees', 2013, 29000000.0),
 ('Alex Rodriguez', 'New York Yankees', 2008, 28000000.0)]

In [8]:
#join query to find the average salary of player's by their team and year

In [9]:
%%sql
SELECT 
p.nameFirst || ' ' || p.nameLast AS fullName,
t.name AS teamName,
s.yearID,
AVG(s.salary) AS avgSalary
FROM Salaries s
JOIN People p ON p.playerID = s.playerID 
JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
GROUP BY teamName, s.yearID ORDER BY avgSalary DESC LIMIT 20 


 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('Alfredo Aceves', 'New York Yankees', 2010, 8253335.56),
 ('Zoilo Almonte', 'New York Yankees', 2014, 8230996.125),
 ('Kevin Brown', 'New York Yankees', 2005, 8011800.653846154),
 ('Jonathan Albaladejo', 'New York Yankees', 2009, 7748045.730769231),
 ('Dustin Ackley', 'New York Yankees', 2016, 7689579.0344827585),
 ('David Adams', 'New York Yankees', 2013, 7483189.870967742),
 ('Brett Anderson', 'Los Angeles Dodgers', 2015, 7441103.448275862),
 ('Carlos Beltran', 'New York Yankees', 2015, 7336274.379310345),
 ('Josh Beckett', 'Los Angeles Dodgers', 2013, 6980068.625),
 ('Luis Ayala', 'New York Yankees', 2011, 6975000.9655172415),
 ('Miguel Cairo', 'New York Yankees', 2006, 6952252.821428572),
 ('Bobby Abreu', 'New York Yankees', 2008, 6929892.966666667),
 ('Al Alburquerque', 'Detroit Tigers', 2015, 6891390.0),
 ('Ehire Adrianza', 'San Francisco Giants', 2016, 6890151.12),
 ('Erisbel Arruebarrena', 'Los Angeles Dodgers', 2014, 6781706.25),
 ('David Aardsma', 'New York Yankees', 2012, 6776630.655172414),
 ('Bobby Abreu', 'New York Yankees', 2007, 6759251.607142857),
 ('Al Alburquerque', 'Detroit Tigers', 2014, 6645891.304347826),
 ('Mike Adams', 'Philadelphia Phillies', 2013, 6533199.576923077),
 ('Matt Barnes', 'Boston Red Sox', 2016, 6501577.9655172415)]

1.4) The average salary of players on the 2005 New York Yankees was $8,011,800.65

## Question 2

In [10]:
# using the inner join to run the query

In [11]:
%%sql
SELECT
    p.nameFirst || ' ' || p.nameLast AS fullName,
    t.name AS teamName,
    s.yearID,
    s.salary
FROM Salaries s
INNER JOIN People p ON p.playerID = s.playerID
INNER JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
ORDER BY s.salary DESC
LIMIT 10;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('Alex Rodriguez', 'New York Yankees', 2009, 33000000.0),
 ('Alex Rodriguez', 'New York Yankees', 2010, 33000000.0),
 ('Clayton Kershaw', 'Los Angeles Dodgers', 2016, 33000000.0),
 ('Clayton Kershaw', 'Los Angeles Dodgers', 2015, 32571000.0),
 ('Alex Rodriguez', 'New York Yankees', 2011, 32000000.0),
 ('Zack Greinke', 'Arizona Diamondbacks', 2016, 31799030.0),
 ('Alex Rodriguez', 'New York Yankees', 2012, 30000000.0),
 ('David Price', 'Boston Red Sox', 2016, 30000000.0),
 ('Alex Rodriguez', 'New York Yankees', 2013, 29000000.0),
 ('Alex Rodriguez', 'New York Yankees', 2008, 28000000.0)]

In [12]:
#calculating the total number of rows

In [13]:
%%sql
SELECT
    COUNT(p.playerID)
FROM Salaries s
INNER JOIN People p ON p.playerID = s.playerID
INNER JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[(26428,)]

In [14]:
# using the left join to run the query

In [15]:
%%sql
SELECT
    p.nameFirst || ' ' || p.nameLast AS fullName,
    t.name AS teamName,
    s.yearID,
    s.salary
FROM Salaries s
LEFT JOIN People p ON p.playerID = s.playerID
LEFT JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
ORDER BY s.salary DESC
LIMIT 10;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('Alex Rodriguez', 'New York Yankees', 2009, 33000000.0),
 ('Alex Rodriguez', 'New York Yankees', 2010, 33000000.0),
 ('Clayton Kershaw', 'Los Angeles Dodgers', 2016, 33000000.0),
 ('Clayton Kershaw', 'Los Angeles Dodgers', 2015, 32571000.0),
 ('Alex Rodriguez', 'New York Yankees', 2011, 32000000.0),
 ('Zack Greinke', 'Arizona Diamondbacks', 2016, 31799030.0),
 ('Alex Rodriguez', 'New York Yankees', 2012, 30000000.0),
 ('David Price', 'Boston Red Sox', 2016, 30000000.0),
 ('Alex Rodriguez', 'New York Yankees', 2013, 29000000.0),
 ('Alex Rodriguez', 'New York Yankees', 2008, 28000000.0)]

In [16]:
#calculating the total number of rows

In [17]:
%%sql
SELECT
    COUNT(p.playerID) AS totalRows
FROM Salaries s
LEFT JOIN People p ON p.playerID = s.playerID
LEFT JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[(26428,)]

2.3) The difference in the results between the two queries is that the inner join query returns only the rows that have matching values between their tables. In our case, the inner join query returned only the rows from the People table where the playerID column matched with the playerID column from the salaries table. Therefore, there are no null values in an inner join. In contrast, the left join returns ALL rows from the table you are joining on the left and then all the matching values of the table you are joining, therefore null values can occur where they are places where the two tables do not have rows containing the same values. 

## Question 3

In [18]:
# using the example query to understand how the HAVING clause works

In [19]:
%%sql
SELECT
    t.name AS teamName,
    s.yearID,
    AVG(s.salary) AS avgSalary
FROM Salaries s
JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
GROUP BY t.name, s.yearID
HAVING AVG(s.salary) > 5000000
ORDER BY avgSalary DESC
LIMIT 10;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('New York Yankees', 2010, 8253335.56),
 ('New York Yankees', 2014, 8230996.125),
 ('New York Yankees', 2005, 8011800.653846154),
 ('New York Yankees', 2009, 7748045.730769231),
 ('New York Yankees', 2016, 7689579.0344827585),
 ('New York Yankees', 2013, 7483189.870967742),
 ('Los Angeles Dodgers', 2015, 7441103.448275862),
 ('New York Yankees', 2015, 7336274.379310345),
 ('Los Angeles Dodgers', 2013, 6980068.625),
 ('New York Yankees', 2011, 6975000.9655172415)]

In [20]:
#writing a query to join the salaries, people and teams table and then group results by year and team and filter only the teams with avgSalary > 7.5 million

In [21]:
%%sql
SELECT 
t.name AS teamName,
s.yearID,
AVG(s.salary) AS avgSalary
FROM Salaries s 
JOIN Teams t ON t.teamID = s.teamID AND t.yearID=s.yearID
GROUP BY t.teamID, s.yearID
HAVING avgSalary > 7500000
ORDER BY avgSalary 

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('New York Yankees', 2016, 7689579.0344827585),
 ('New York Yankees', 2009, 7748045.730769231),
 ('New York Yankees', 2005, 8011800.653846154),
 ('New York Yankees', 2014, 8230996.125),
 ('New York Yankees', 2010, 8253335.56)]

3.3) Only 5 teams meet this criteria, the 2005, 2009, 2010, 2014, and 2016 New York Yankees.

## Question 4

In [22]:
# using the multiple condition HAVING clause query example to examine functionality

In [23]:
%%sql
SELECT
    t.name AS teamName,
    s.yearID,
    AVG(s.salary) AS avgSalary,
    COUNT(s.playerID) AS numPlayers
FROM Salaries s
JOIN Teams t ON t.teamID = s.teamID AND t.yearID = s.yearID
GROUP BY t.name, s.yearID
HAVING AVG(s.salary) > 5000000 AND COUNT(s.playerID) > 20
ORDER BY avgSalary DESC
LIMIT 10;

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('New York Yankees', 2010, 8253335.56, 25),
 ('New York Yankees', 2014, 8230996.125, 24),
 ('New York Yankees', 2005, 8011800.653846154, 26),
 ('New York Yankees', 2009, 7748045.730769231, 26),
 ('New York Yankees', 2016, 7689579.0344827585, 29),
 ('New York Yankees', 2013, 7483189.870967742, 31),
 ('Los Angeles Dodgers', 2015, 7441103.448275862, 29),
 ('New York Yankees', 2015, 7336274.379310345, 29),
 ('Los Angeles Dodgers', 2013, 6980068.625, 32),
 ('New York Yankees', 2011, 6975000.9655172415, 29)]

In [24]:
# query to find teams with average salary greater than 6.8, more than 20 players and the year being between 2010 and 2015 

In [25]:
%%sql
SELECT 
t.name AS teamName,
s.yearID,
AVG(s.salary) AS avgSalary,
COUNT(s.playerID) AS totalPlayers
FROM Salaries s
JOIN Teams t ON t.teamID = s.TeamID AND t.yearID = s.yearID
WHERE s.yearID BETWEEN 2010 AND 2015
GROUP BY t.name, s.yearID
HAVING avgSalary > 6800000 AND totalPlayers > 20
ORDER BY avgSalary DESC

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('New York Yankees', 2010, 8253335.56, 25),
 ('New York Yankees', 2014, 8230996.125, 24),
 ('New York Yankees', 2013, 7483189.870967742, 31),
 ('Los Angeles Dodgers', 2015, 7441103.448275862, 29),
 ('New York Yankees', 2015, 7336274.379310345, 29),
 ('Los Angeles Dodgers', 2013, 6980068.625, 32),
 ('New York Yankees', 2011, 6975000.9655172415, 29),
 ('Detroit Tigers', 2015, 6891390.0, 25)]

4.4) There are 8 teams that fit this criteria. The 2010, 2011, 2013, 2014, and 2015 New York Yankees, the 2013 and 2015 Los Angeles Dodgers and the 2015 Detroit Tigers

## Question 5

In [26]:
# creating the payroll aggregation for each team in each year

In [27]:
%%sql 
WITH payroll AS (
    SELECT
        teamID,
        yearID,
        SUM(salary) AS totalPayroll
    FROM Salaries
    GROUP BY teamID, yearID
)
SELECT * FROM payroll WHERE totalPayroll > 150000000 LIMIT 10

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('BAL', 2016, 161863456.0),
 ('BOS', 2010, 162447333.0),
 ('BOS', 2011, 161762475.0),
 ('BOS', 2012, 173186617.0),
 ('BOS', 2013, 151530000.0),
 ('BOS', 2015, 181103400.0),
 ('BOS', 2016, 188545761.0),
 ('CHN', 2016, 154067668.0),
 ('DET', 2014, 152855500.0),
 ('DET', 2015, 172284750.0)]

In [28]:
#finding the max number of home runs for each player in each year 

In [29]:
%%sql
WITH maxHR AS (
    SELECT 
        teamID,
        playerID,
        yearID,
        SUM(HR) AS totalHR
    FROM Batting
    GROUP BY playerID, teamID, yearID
)

SELECT teamID,playerID, yearID, MAX(totalHR) AS maxHR 
FROM maxHR 
GROUP BY teamID, yearID
HAVING maxHR >= 30
ORDER BY maxHR DESC LIMIT 20

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('SFN', 'bondsba01', 2001, 73),
 ('SLN', 'mcgwima01', 1998, 70),
 ('CHN', 'sosasa01', 1998, 66),
 ('SLN', 'mcgwima01', 1999, 65),
 ('CHN', 'sosasa01', 2001, 64),
 ('CHN', 'sosasa01', 1999, 63),
 ('NYA', 'judgeaa01', 2022, 62),
 ('NYA', 'marisro01', 1961, 61),
 ('NYA', 'ruthba01', 1927, 60),
 ('MIA', 'stantmi03', 2017, 59),
 ('NYA', 'ruthba01', 1921, 59),
 ('DET', 'greenha01', 1938, 58),
 ('PHA', 'foxxji01', 1932, 58),
 ('PHI', 'howarry01', 2006, 58),
 ('ARI', 'gonzalu01', 2001, 57),
 ('TEX', 'rodrial01', 2002, 57),
 ('CHN', 'wilsoha01', 1930, 56),
 ('SEA', 'griffke02', 1997, 56),
 ('SEA', 'griffke02', 1998, 56),
 ('ATL', 'olsonma02', 2023, 54)]

In [30]:
#joining the two aggregates and filtering to keep only qualifying team-seasons

In [31]:
%%sql
WITH payroll AS (
    SELECT
        teamID,
        yearID,
        SUM(salary) AS totalPayroll
    FROM Salaries
    GROUP BY teamID, yearID
),
maxHR AS (
    SELECT 
        teamID,
        playerID,
        yearID,
        SUM(HR) AS totalHR
    FROM Batting
    GROUP BY playerID, teamID, yearID
)
SELECT 
    p.teamID,
    p.yearID,
    p.totalPayroll,
    m.playerID,
    m.totalHR
FROM payroll p
JOIN maxHR m ON m.teamID = p.teamID AND m.yearID = p.yearID
GROUP BY p.teamID, p.yearID
HAVING MAX(m.totalHR) >= 30 AND p.totalPayroll > 150000000
ORDER BY p.totalPayroll LIMIT 5

 * sqlite:////anvil/projects/tdm/data/lahman/lahman.db
Done.


[('BOS', 2013, 151530000.0, 'ortizda01', 30),
 ('NYA', 2003, 152749814.0, 'giambja01', 41),
 ('DET', 2014, 152855500.0, 'martivi01', 32),
 ('CHN', 2016, 154067668.0, 'bryankr01', 39),
 ('LAA', 2012, 154485166.0, 'trumbma01', 32)]

5.2) The top 5 cheapest teams were the 2013 Boston Red Sox, 2003 New York Yankees, 2014 Detroit Tigers, 2016 Chicago Cubs, and the 2012 Los Angeles Angels

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.